hibernate-core/src/main/java/org/hibernate/query/sqm/sql/BaseSqmToSqlAstConverter.java

```diff
@@ -6931,9 +6931,9 @@ public LikePredicate visitLikePredicate(SqmLikePredicate predicate) {
 		return new LikePredicate(
 				visitWithInferredType( predicate.getMatchExpression(), predicate.getPattern() ),
 				visitWithInferredType( predicate.getPattern(), predicate.getMatchExpression() ),
-				predicate.getEscapeCharacter() == null
-						? null
-						: visitWithInferredType( predicate.getEscapeCharacter(), predicate.getMatchExpression() ),
+				predicate.getEscapeCharacter() == null ?
+						null :
+						(Expression) predicate.getEscapeCharacter().accept( this ),
```

Comment:
I think we should push the `StandardBasicTypes.CHARACTER` type on the inference stack when visiting the escape character i.e. push `creationContext.getSessionFactory().getTypeConfiguration().getBasicTypeRegistry().resolve(StandardBasicTypes.CHARACTER)`, then `predicate.getEscapeCharacter().accept( this )` and finally pop again.

Reply:
Done this with `visitWithInferredType( predicate.getEscapeCharacter(), () -> basicType( Character.class ) )`.